maven/lib/dependabot/maven/file_parser.rb

```diff
@@ -386,7 +410,78 @@ def internal_dependency_names
       def check_required_files
         raise "No pom.xml!" unless get_original_file("pom.xml")
       end
+
+      # Merge dependency scan requirements with file parsing requirements.
+      # Since dependency scan evaluates properties, we need to combine results with XML parsing,
+      # so we know when certain requirement not a literal value and can differentiate transitive dependencies
+      # from direct dependencies.
+      sig do
+        params(requirements: T::Array[T::Hash[Symbol, T.untyped]]).returns(T::Array[T::Hash[Symbol, T.untyped]])
+      end
+      def merge_requirements(requirements)
+        return requirements if requirements.length <= 1
+
+        merged = []
+        used_indices = Set.new
+        requirements.each_with_index do |dep_scan_req, i|
+          next if used_indices.include?(i) || dep_scan_req.dig(:metadata, :pom_file).nil?
+
+          # Look for another requirement where pom_file matches property_source
+          match_index = requirements.find_index.with_index do |parsing_req, j|
+            j > i &&
+              !used_indices.include?(j) &&
+              dep_scan_req.dig(:metadata, :pom_file) == parsing_req.fetch(:file)
+          end
+
+          if match_index
+            parsing_req = T.must(requirements[match_index])
+
+            # Merge the two requirements
+            # We prefer file and requirement properties from parsed requirements,
+            # because they include correct file and not evaluated property value.
+            merged_req = {
+              requirement: parsing_req[:requirement],
+              file: parsing_req[:file],
+              groups: (Array(dep_scan_req[:groups]) + Array(parsing_req[:groups])).uniq,
```

Comment:
nit: you can use the splat (`*`) operator here:


```suggestion
              groups: [*dep_scan_req[:groups], *parsing_req[:groups]].uniq.compact,
```